src/Confluent.SchemaRegistry/Rest/RestService.cs

```diff
@@ -0,0 +1,302 @@
+                    // 4xx errors should not be retried (these are conclusive)
+                    if (response.StatusCode == HttpStatusCode.InternalServerError)
+                    {
+                        if (!firstError)
+                        {
+                            aggregatedErrorMessage += "; ";
+                        }
+                        else 
+                        {
+                            firstError = false;
+                        }
+
+                        string message = "";
+                        int errorCode = -1;
+                        try
+                        {
+                            var errorObject = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+                            message = errorObject.Value<string>("message");
+                            errorCode = errorObject.Value<int>("error_code");
+                        }
+                        catch
+                        {
+                            aggregatedErrorMessage += $"[{clients[i].BaseAddress}] {response.StatusCode}";
+                        }
+
+                        aggregatedErrorMessage += $"[{clients[i].BaseAddress}] {response.StatusCode} {errorCode} {message}";
+                    }
+                    else 
+                    {
+                        gotOkAnswer = true;
+                    }
+                }
+                catch (HttpRequestException e)
+                {
+                    if (!firstError)
+                    {
+                        aggregatedErrorMessage += "; ";
+                    }
+                    else 
+                    {
+                        firstError = false;
+                    }
+
+                    aggregatedErrorMessage += $"[{clients[i].BaseAddress}] HttpRequestException: {e.Message}";
+                }
+            }
+
+            if (!gotOkAnswer)
+            {
+                // All SR calls failed, return an exception containing information about each failure.
+                // TODO: Allow non-empty aggregatedErrorMessage to be logged even when gotOkAnswer == true.
+                throw new HttpRequestException(aggregatedErrorMessage);
+            }
+
+            if (!response.IsSuccessStatusCode)
+            {
+                // 4xx errors
+                var errorObject = JObject.Parse(await response.Content.ReadAsStringAsync().ConfigureAwait(false));
+                var message = errorObject.Value<string>("message");
+                var errorCode = errorObject.Value<int>("error_code");
+                throw new SchemaRegistryException(message, response.StatusCode, errorCode);
+            }
+
+            // if we had success, set last client used so we start with this next time.
+            lock (lastClientUsedLock)
+            {
+                this.lastClientUsed = (this.lastClientUsed + 1) % clients.Count;
```

Comment:
shouldn't this be `lastClientUsed = i` ?

Reply:
whoa, that's due to a bad refactor. refactored again, please check this logic again.